src/databricks/labs/ucx/framework/history.py

```diff
@@ -0,0 +1,155 @@
+
+    failures: list[str]
+    """The list of problems associated with the object that this inventory record covers."""
+
+    owner: str
+    """The identity of the account that created this inventory record."""
+
+
+class DataclassInstance(Protocol):
+    __dataclass_fields__: ClassVar[dict]
+
+
+Record = TypeVar("Record", bound=DataclassInstance)
+
+
+class HistoryLog:
+    __slots__ = ("_ws", "_backend", "_run_id", "_catalog", "_schema", "_table")
+
+    def __init__(
+        self,
+        ws: WorkspaceClient,
+        backend: SqlBackend,
+        run_id: int,
+        catalog: str,
+        schema: str,
+        table: str,
+    ) -> None:
+        self._ws = ws
+        self._backend = backend
+        self._run_id = run_id
+        self._catalog = catalog
+        self._schema = schema
+        self._table = table
+
+    @property
+    def full_name(self) -> str:
+        return f"{self._catalog}.{self._schema}.{self._table}"
+
+    def _append_history_snapshot(self, object_type: str, snapshot: list[HistoricalRecord]) -> None:
+        logger.debug(f"[{self.full_name}] appending {len(snapshot)} new records for {object_type}")
+        # Concurrent writes do not need to be handled here; appends cannot conflict.
+        # TODO: Although documented as conflict-free, verify that this is truly is the case.
+        self._backend.save_table(self.full_name, snapshot, HistoricalRecord, mode="append")
+
+    class Appender:
+        __slots__ = ("_ws", "_object_type", "_object_type_version", "_key_from", "_run_id", "_persist")
+
+        def __init__(
+            self,
+            ws: WorkspaceClient,
+            run_id: int,
+            klass: type[Record],
+            key_from: Callable[[Record], str],
+            persist: Callable[[str, list[HistoricalRecord]], None],
+        ) -> None:
+            self._ws = ws
+            self._run_id = run_id
+            self._object_type = klass.__name__
+            # Versioning support: if the dataclass has a _ucx_version class attribute that is the current version.
+            self._object_type_version = getattr(klass, "_ucx_version") if hasattr(klass, "_ucx_version") else 0
+            self._key_from = key_from
+            self._persist = persist
+
+        @cached_property
+        def _workspace_id(self) -> int:
+            return self._ws.get_workspace_id()
+
+        @cached_property
+        def _owner(self) -> str:
+            current_user = self._ws.current_user.me()
+            owner = current_user.user_name or current_user.id
+            assert owner
+            return owner
+
+        def append_snapshot(self, records: Sequence[Record], *, run_start_time: dt.datetime) -> None:
+            # Equivalent entropy to a type-4 UUID.
+            snapshot_id = int.from_bytes(os.urandom(16), byteorder="big")
+            historical_records = [
+                self._inventory_record_to_historical(record, snapshot_id=snapshot_id, run_start_time=run_start_time)
```

Comment:
@JCZuurmond in the `parse_logs` task, create the `WorkflowRuns` entities, where you record run start time and run end time. otherwise the writer side gets too complicated. `parse_logs` is guaranteed to be executed at the end of every job run, so it's a good place to do so. or add one more task explicitly, which is a bit more explicit.

Reply:
https://github.com/databrickslabs/ucx/pull/2754